aea/helpers/dependency_tree.py

```diff
@@ -117,18 +118,61 @@ def get_all_dependencies(item_config: Dict) -> List[PackageId]:
         )
 
     @classmethod
-    def resolve_tree(
-        cls, dependency_list: Dict[PackageId, List[PackageId]], tree: Dict
+    def resolve_tree(cls, dependency_list: Dict[PackageId, List[PackageId]]) -> Dict:
+        """
+        Resolve dependency tree.
+
+        :param dependency_list: the adjacency list of the dependency graph
+        :return: the dependency tree
+        """
+        tree: Dict[PackageId, Dict] = {root_node: {} for root_node in dependency_list}
+        # auxiliary variables to keep track during recursive calls :
+        #  visited: a set for fast checking of already visited nodes in the current path of DFS
+        #  stack:   the current stack of nodes in the depth-first visiting of nodes
+        visited: Set[PackageId] = set()
+        stack: List[PackageId] = []
+        cls._resolve_tree_aux(dependency_list, tree, visited, stack)
+
+        # _resolve_tree_aux did side-effect on the 'tree' object
+        return tree
+
+    @classmethod
+    def _resolve_tree_aux(
+        cls,
+        dependency_list: Dict[PackageId, List[PackageId]],
+        tree: Dict,
+        visited: Set[PackageId],
+        stack: List[PackageId],
     ) -> None:
-        """Resolve dependency tree"""
+        """
+        Resolve dependency tree (auxiliary method of resolve_tree).
+
+        IMPORTANT: This method does side-effect on the parameter 'tree', 'visited' and 'stack'.
 
+        :param dependency_list: the adjacency list of hte dependency graph
+        :param tree: the dependency tree
+        :param visited: the set of visited nodes in the current path
+        :param stack: the current path stored in a LIFO data structure
+        """
         for root_package in tree:
+            if root_package in visited:
+                # cycle found; raise error
+                cls._raise_circular_dependency_error(root_package, stack)
+                return
+
+            # add current package to the stack - needed to check if there are circular dependencies
+            visited.add(root_package)
+            stack.append(root_package)
+
+            # build the root of the subtree according to the dependency list
             root_dependencies: List[PackageId] = dependency_list.get(root_package, [])
-            for package in root_dependencies:
-                tree[root_package][package] = {
-                    p: {} for p in cast(list, dependency_list.get(package))
-                }
-                cls.resolve_tree(dependency_list, tree[root_package][package])
```

Comment:
this for-loop implemented a second recursive step in the same call of the recursive function. Removed as I thought could possibly improve readability and code structure (in particular, note that if remained as before I would have had to append-to/pop-from the stack before/after the recursive call in the inner for-loop, which looks fishy). Unless I back the wrong horse, the behaviour of the method should be equivalent.